### PR TITLE
fix: Communicate errors when resolve function can not operate

### PR DESCRIPTION
**Pull request by RafalKorepta**
#### Submitter checklist

- [ ] Change is code complete and matches issue description.
- [X] Change is covered by existing or new tests.

#### Description

The function resolve should handle edge cases like niled pointer receiver and niled user argument by providing information to the user.

#### Special notes for the reviewer(s)

The messages that resolve function returns are not very informative for the user of jx command-line tool, but developers could quickly identify a problem when the messages appears.

### CODE REVIEW

**Comment by hferentschik** [comment]
@RafalKorepta, thanks for the pull request. Any chance you could also create a corresponding issue? Some more context would be ideal in terms when these edge cases appeared and what the exception was in this case.

File: pkg/cmd/preview/preview.go [hunk 1]
```diff
@@ -286,24 +286,21 @@ func (o *PreviewOptions) Run() error {
...
-						log.Logger().Warnf("Unable to update user %s from %s because %v", prAuthor.Name, o.PullRequestName, err)
-					}
-					if author != nil {
-						user = &v1.UserSpec{
-							Username: author.Spec.Login,
-							Name:     author.Spec.Name,
-							ImageURL: author.Spec.AvatarURL,
-							LinkURL:  author.Spec.URL,
-						}
+				author, err := resolver.Resolve(pullRequest.Author)
```
Comment (hferentschik) [inline pkg/cmd/preview/preview.go]: Maybe the intention was to ignore a non-set author, as opposed to failing the command. 
Was there another error occurring with the preview command when the nil author was nil?

File: pkg/users/git.go [hunk 2]
```diff
@@ -60,8 +60,11 @@ func (r *GitUserResolver) GitUserSliceAsUserDetailsSlice(users []gits.GitUser) (
 // * making a call to the gitProvider
 // as often user info is not complete in a git response
 func (r *GitUserResolver) Resolve(user *gits.GitUser) (*jenkinsv1.User, error) {
-	if r == nil || user == nil {
-		return nil, nil
+	if r == nil {
+		return nil, errors.New("unable to resolve user on niled receiver")
+	}
+	if user == nil {
+		return nil, errors.New("unable to resolve when user is niled")
```
Comment (hferentschik) [inline pkg/users/git.go]: I think it should be "unable to resolve when user is nil".

File: pkg/users/git.go [hunk 3]
```diff
@@ -60,8 +60,11 @@ func (r *GitUserResolver) GitUserSliceAsUserDetailsSlice(users []gits.GitUser) (
 // * making a call to the gitProvider
 // as often user info is not complete in a git response
 func (r *GitUserResolver) Resolve(user *gits.GitUser) (*jenkinsv1.User, error) {
-	if r == nil || user == nil {
-		return nil, nil
+	if r == nil {
+		return nil, errors.New("unable to resolve user on niled receiver")
```
Comment (hferentschik) [inline pkg/users/git.go]: I think it should be "unable to resolve user on nil receiver". I don't think "niled" is a thing. At least I have not seen it used anywhere before.

Comment (RafalKorepta) [reply pkg/users/git.go]: You are right.

File: pkg/users/git.go (hunk 2 above)
Comment (RafalKorepta) [reply pkg/users/git.go]: Agree.

File: pkg/cmd/preview/preview.go (hunk 1 above)
Comment (RafalKorepta) [reply pkg/cmd/preview/preview.go]: I didn't think of such use case. In that case I should revert the preview.go change entirely.

**Comment by RafalKorepta** [comment]
@hferentschik I didn't find any problem when I used preview command. On my debugging session, I realise that I could improve unit test for `GitUserResolve.Resolve()`. I will leave you to decide if this PR is worth merging.

**Comment by hferentschik** [comment]
> I will leave you to decide if this PR is worth merging.

@RafalKorepta, thanks again for the PR. I will close it for now. It is not only _preview_ which relies on this _Resolve_, but also changelog generation. I'd also wish that the code would be a bit better, especially in terms of what it reports to the user, but by merging this change we basically change the semantics of the code. For example, the changelog generation should not error out if a single user cannot be resolved.

Eventually, I'd like to revisit this code, but I think it needs to go hand in hand with a bigger refactoring.
